### ISSUE TITLE
Use "bt" for switch select from bitmask?

### ISSUE THREAD

**Issue by zmodem** [llvm/llvm-project#38522]
|  |  |
| --- | --- |
| Bugzilla Link | [39174](https://llvm.org/bz39174) |
| Version | trunk |
| OS | Linux |
| CC | @topperc,@LebedevRI,@RKSimon,@rotateright |

## Extended Description 
For example:
```
bool f(int x) {
    switch (x) {
    case 0: return true;
    case 1: return true;
    case 2: return false;
    case 3: return false;
    case 4: return true;
    }
}
```
SimplifyCFG will generate a "table lookup" from a bitmask, and on X86_64 we end up with:
```
f(int):
# %bb.0:
        mov     al, 19
        mov     ecx, edi
        shr     al, cl
        and     al, 1
        ret
```

I wonder if we could get tighter code with the BT instruction.

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D122891

**Comment by RKSimon** [comment]
Fixed by b8652fbcbbde830ec2f48ff165e5706fb607d02b

https://simd.godbolt.org/z/9c1YjszrY
```
f1:
  movl $19, %eax
  btl %edi, %eax
  setb %al
  retq
```
